File: problems/CSES/Dynamic_Programming/counting_numbers/counting_numbers.py
import sys
from functools import lru_cache
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def count_valid(x: int) -> int:
    if x < 0:
        return 0

    digits = str(x)
    n = len(digits)

    @lru_cache(maxsize=None)
    def dp(
        pos: int,
        previous: int,
        tight: bool,
        started: bool
    ) -> int:
        if pos == n:
            return 1

        limit = int(digits[pos]) if tight else 9
        res = 0

        for digit in range(limit + 1):
            next_tight = tight and digit == limit

            if not started and digit == 0:
                res += dp(
                    pos + 1,
                    10,
                    next_tight,
                    False
                )
                continue

            if started and digit == previous:
                continue

            res += dp(
                pos + 1,
                digit,
                next_tight,
                True
            )

        return res

    return dp(0, 10, True, False)

def main():
    a, b = map(int, input().split())

    print(count_valid(b) - count_valid(a - 1))

main()

File: problems/CSES/Dynamic_Programming/counting_numbers/counting_numbers.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

string digits;
ll memo[20][11][2][2];
bool visited[20][11][2][2];

ll dp(int pos, int previous, bool tight, bool started) {
    if (pos == digits.size()) {
        return 1;
    }

    if (visited[pos][previous][tight][started]) {
        return memo[pos][previous][tight][started];
    }

    visited[pos][previous][tight][started] = true;

    int currentDigit = digits[pos] - '0';
    int limit = tight ? currentDigit : 9;

    ll res = 0;

    for (int digit = 0; digit <= limit; digit++) {
        bool nextTight = tight && (digit == currentDigit);

        if (!started && digit == 0) {
            res += dp(
                pos + 1,
                10,
                nextTight,
                false
            );

            continue;
        }

        if (started && digit == previous) {
            continue;
        }

        res += dp(
            pos + 1,
            digit,
            nextTight,
            true
        );
    }

    return memo[pos][previous][tight][started] = res;
}

ll count_valid(ll x) {
    if (x < 0) {
        return 0;
    }

    digits = to_string(x);

    memset(visited, false, sizeof(visited));

    return dp(0, 10, true, false);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    ll a, b;
    cin >> a >> b;

    cout << count_valid(b) - count_valid(a - 1) << '\n';

    return 0;
}
